src/datadog/w3c_propagation.cpp

```diff
@@ -0,0 +1,396 @@
+    return;
+  }
+
+  auto& [datadog_value, other_entries] = *maybe_parsed;
+  if (!other_entries.empty()) {
+    result.additional_w3c_tracestate = std::move(other_entries);
+  }
+
+  parse_datadog_tracestate(result, datadog_value);
+}
+
+}  // namespace
+
+Expected<ExtractedData> extract_w3c(
+    const DictReader& headers,
+    std::unordered_map<std::string, std::string>& span_tags, Logger&) {
+  ExtractedData result;
+
+  if (auto error_tag_value = extract_traceparent(result, headers)) {
+    span_tags[tags::internal::w3c_extraction_error] =
+        std::move(*error_tag_value);
+    return ExtractedData{};
+  }
+
+  // If we didn't get a trace ID from traceparent, don't bother with
+  // tracestate.
+  if (!result.trace_id) {
+    return result;
+  }
+
+  extract_tracestate(result, headers);
+
+  return result;
+}
+
+std::string encode_traceparent(
+    std::uint64_t trace_id, const Optional<std::string>& full_w3c_trace_id_hex,
+    std::uint64_t span_id, int sampling_priority) {
+  std::string result;
+  // version
+  result += "00-";
+
+  // trace ID
+  if (full_w3c_trace_id_hex) {
+    result += *full_w3c_trace_id_hex;
+  } else {
+    auto hexed = hex(trace_id);
+    result.append(16 + (16 - hexed.size()), '0');  // leading zeroes
+    result += hexed;
+  }
+  result += '-';
+
+  // span ID
+  auto hexed = hex(span_id);
+  result.append(16 - hexed.size(), '0');  // leading zeroes
+  result += hexed;
+  result += '-';
+
+  // flags
+  result += sampling_priority > 0 ? "01" : "00";
+
+  return result;
+}
+
+std::string encode_datadog_tracestate(
+    int sampling_priority, const Optional<std::string>& origin,
+    const std::vector<std::pair<std::string, std::string>>& trace_tags,
+    const Optional<std::string>& additional_datadog_w3c_tracestate) {
+  std::string result = "dd=s:";
+  result += std::to_string(sampling_priority);
+
+  if (origin) {
+    result += ";o:";
+    result += *origin;
+    std::replace_if(result.end() - origin->size(), result.end(),
+                    verboten(0x20, 0x7e, ",;="), '_');
+  }
+
+  for (const auto& [key, value] : trace_tags) {
+    const StringView prefix = "_dd.p.";
+    if (!starts_with(key, prefix)) {
+      continue;
+    }
+
+    // `key` is "_dd.p.<name>", but we want "t.<name>".
+    result += ";t.";
+    result.append(key, prefix.size());
+    std::replace_if(result.end() - (key.size() - prefix.size()), result.end(),
+                    verboten(0x20, 0x7e, " ,;="), '_');
+
+    result += ':';
+    result += value;
+    std::replace_if(result.end() - value.size(), result.end(),
+                    verboten(0x20, 0x7e, ",;~"), '_');
+    // `value` might contain equal signs ("="), which is reserved in tracestate.
+    // Replace them with tildes ("~").
+    std::replace(result.end() - value.size(), result.end(), '=', '~');
+  }
+
+  if (additional_datadog_w3c_tracestate) {
+    result += ';';
+    result += *additional_datadog_w3c_tracestate;
+  }
+
+  const std::size_t max_size = 256;
```

Comment:
Is this an arbitrary size or is it defined somewhere?

Reply:
It's defined in our W3C propagation RFC. It's not configurable because it is derived from the W3C definition itself.